### ISSUE TITLE
[java] CloneMethodMustImplementCloneable triggers for interfaces

### ISSUE THREAD

**Issue by huxi** [pmd/pmd#1005]
**Affects PMD Version:** 
6.2.0

**Rule:**
category/java/errorprone.xml/CloneMethodMustImplementCloneable

**Description:**
The code below did not trigger this rule before PMD 6.2.0.

The `clone()` method is defined in the interface so the return value is of type `Foo` instead of `Object`.

**Code Sample demonstrating the issue:**

```java
public interface Foo
    extends Serializable, Cloneable {
    Foo clone() throws CloneNotSupportedException;
}
```

**Running PMD through:** *Gradle*

**Comment by adangel** [comment]
Fixed via #2899 for PMD 7.

**Comment by Fernal73** [comment]
Also occuring for class , not just interface.
```java
package ds;

import java.util.concurrent.Callable;

final class BubbleTask implements Cloneable, Callable<Void> {
  long[] a;
  int i;
  AbstractBrickSort sorter;

  @SuppressWarnings("PMD.ArrayIsStoredDirectly")
  BubbleTask(AbstractBrickSort sorter, long[] a, int i) {
    this.sorter = sorter;
    this.a = a;
    this.i = i;
  }

  @Override
  public BubbleTask clone() {
    try {
    return (BubbleTask)super.clone();
    } catch (CloneNotSupportedException cnse) {
     throw new AssertionError("Shouldn't get here.." + cnse.getMessage(),
         cnse);
    }
  }

  @Override
  public Void call() {
    sorter.bubble(a, i);
    return null;
  }

  public static BubbleTask createCopy(BubbleTask bt) {
    return (BubbleTask)bt.clone();
  }
}
```

Is this fixed in PMD 7.0.0 as well?

**Comment by Fernal73** [comment]
Apparently, the above is fixed on compilation , probably to do with auxclasspath.

But then there's CloneThrowsCloneNotSupportedException for this:

```
package ds;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.PriorityQueue;

/*
 * Compress or expand a binary input stream using the Huffman algorithm.
 *
 * <p>****************************************************************************
 * </p>
 */

/**
 * The {@code HuffmanBase} class provides methods for a binary input using Huffman codes over the
 * 8-bit extended ASCII alphabet.
 *
 * <p>For additional documentation, see <a
 * href="https://algs4.cs.princeton.edu/55compression">Section 5.5</a> of <i>Algorithms, 4th
 * Edition</i> by Robert Sedgewick and Kevin Wayne.
 *
 * @author Robert Sedgewick
 * @author Kevin Wayne
 */
@SuppressWarnings({
  "checkstyle:AbstractClassName",
  "PMD.AbstractClassWithoutAbstractMethod",
  "nullness"
})
public abstract class HuffmanBase {

  // alphabet size of extended ASCII
  static final int R = 256;

  static final char NULL_CHARACTER = '\0';
  static final char ZERO_CHARACTER = '0';
  static final char ONE_CHARACTER = '1';

  protected final BinaryInputStream bis;
  protected final BinaryOutputStream bos;

  public HuffmanBase(File input, File output) throws IOException {
    this.bis = new BinaryInputStream(Files.newInputStream(input.toPath()));
    this.bos = new BinaryOutputStream(Files.newOutputStream(output.toPath()));
  }

  // build the Huffman trie given frequencies
  protected Node buildTrie(int... freq) {

    // initialize priority queue with singleton trees
    PriorityQueue<Node> pq = new PriorityQueue<>();
    Node node = new Node('a', 0, null, null);
    for (char c = 0; c < R; c++)
      if (freq[c] > 0) {
        Node newNode = node.clone();
        newNode.ch = c;
        newNode.freq = freq[c];
        pq.offer(newNode);
      }

    // merge two smallest trees
    while (pq.size() > 1) {
      Node left = pq.poll();
      Node right = pq.poll();
      Node parent = node.clone();
      parent.left = left;
      parent.right = right;
      parent.ch = NULL_CHARACTER;
      parent.freq = left.freq + right.freq;
      pq.offer(parent);
    }
    return pq.poll();
  }

  // write bitstring-encoded trie to standard output
  protected void writeTrie(Node x) {
    if (x.isLeaf()) {
      bos.write(true);
      bos.write(x.ch, 8);
      return;
    }
    bos.write(false);
    writeTrie(x.left);
    writeTrie(x.right);
  }

  // make a lookup table from symbols and their encodings
  protected void buildCode(String[] st, Node x, String s) {
    if (x.isLeaf()) st[x.ch] = s;
    else {
      buildCode(st, x.left, s + '0');
      buildCode(st, x.right, s + '1');
    }
  }

  protected Node readTrie() {
    boolean isLeaf = bis.readBoolean();
    if (isLeaf) {
      char ch = bis.readChar();
      return new Node(ch, -1, null, null);
    } else return new Node(NULL_CHARACTER, -1, readTrie(), readTrie());
  }

  // Huffman trie node
  @SuppressWarnings("nullness")
  static class Node implements Cloneable, Comparable<Node> {
    char ch;
    int freq;
    Node left;
    Node right;

    Node(char ch, int freq, Node left, Node right) {
      this.ch = ch;
      this.freq = freq;
      this.left = left;
      this.right = right;
    }

    // is the node a leaf node?
    boolean isLeaf() {
      assert left == null && right == null || left != null && right != null;
      return left == null && right == null;
    }

    // compare, based on frequency
    @Override
    public int compareTo(Node that) {
      return this.freq - that.freq;
    }

    @Override
    public Node clone() {
      try {
        return (Node) super.clone();
      } catch (CloneNotSupportedException cnse) {
        throw new AssertionError("Shouldn't get here..." + cnse.getMessage(), cnse);
      }
    }
  }
}

```

**Comment by Fernal73** [comment]
This is caused by the class not being declared final. Then, the rule does not flag the method.

Better documentation would help as in [NoClone](https://checkstyle.sourceforge.io/config_coding.html#NoClone).

**Comment by oowekyala** [comment]
@Fernal73 Please minimise the code samples you post here. Also this issue is not about CloneThrowsCloneNotSupportedException but a different rule. Create another ticket if you like

**Comment by Fernal73** [comment]
Thanks, @oowekyala . As far as I can see, that won't be needed.
